pkgs/build-support/node/build-npm-package/default.nix

```diff
@@ -37,12 +37,45 @@ let
     hash = npmDepsHash;
   };
 
+  patchNode =
+    let
+      npmNodeModules = "lib/node_modules/npm/node_modules";
+      setPath = "${npmNodeModules}/@npmcli/run-script/lib/set-path.js";
+    in
+    nodejs: stdenv.mkDerivation {
+      pname = "nodejs-gyp-patched";
+      inherit (nodejs) version;
+
+      src = nodejs;
+
+      patchPhase = ''
+        runHook prePatch
+
+        sed -i "s|const pathArr = \[\]|const pathArr = \[nodeGypPath\]; pathArr.push(PATH)|" ${setPath}
+        sed -i "/pathArr.push(nodeGypPath, PATH)/d" ${setPath}
+
+        sed -i "s|node-gyp rebuild|$out/${npmNodeModules}/node-gyp/bin/node-gyp.js rebuild|" ${npmNodeModules}/@npmcli/node-gyp/lib/index.js
```

Comment:
TODO: explain what all of these do, the quirks associated, etc.

Until I write these comments, I leave you with this as a sneak peek:
<img width="852" alt="image" src="https://user-images.githubusercontent.com/78392041/206611020-a57290e8-ac3d-4400-bac0-be08b3a0bf2c.png">

Reply:
For that first line, is there any reason you can't just `sed -i "s|const pathArr = \[\]|const pathArr = \[nodeGypPath, PATH\]|" ${setPath}`?

Also would `substituteInPlace` be more readable than `sed` since it doesn't look like we're actually intending to use regex meta chars? Or do you think even a patch file would be stable enough to use?

Also one more thought:

Given `process.env` doesn't require importing anything, you can probably adjust the patching to depend on whether `NIX_NODE_GYP_PATH` is defined and use its value like:

```
substituteInPlace --replace 'const pathArr = []' 'const pathArr = process.env.NIX_NODE_GYP_PATH ? [nodeGypPath, PATH] : []'
substituteInPlace --replace 'pathArr.push(nodeGypPath, PATH)' '(process.env.NIX_NODE_GYP_PATH || pathArr.push(nogeGypPath, PATH))' ${setPath}
substituteInPlace --replace "'node-gyp rebuild'" '`${process.env.NIX_NODE_GYP_PATH || "node-gyp"} rebuild`' ${npmNodeModules}/@npmcli/node-gyp/lib/index.js
```

Then you would be able to apply this patch directly at the nodejs derivation (making it easier to use a patch file) and just set `NIX_NODE_GYP_PATH` from within `npmHooks.npmConfigHook` or something

Reply:
> Or do you think even a patch file would be stable enough to use?

Using a patch file is my goal, but I need to look into if it would be stable enough.

Reply:
This one has only seen one change that would affect a patch with 3 line context, and it was to add a comma: https://github.com/npm/cli/blame/latest/node_modules/%40npmcli/node-gyp/lib/index.js

This one has also only had one recent change that would directly affect a patch (that I see) but they've been refactoring it and such between npm releases, so it is more likely to need minor changes to a patch with 3 lines context: https://github.com/npm/cli/blame/latest/node_modules/%40npmcli/run-script/lib/set-path.js